Wrong build order with nested dependencies

Hi, first of all, thank you for working on such a great tool for dependency-management :+1: 

I have encountered a weird issue with nested dependencies and I am not sure if Carthage works as intended here.

I have the following dependency graph (A requires B, B requires C, ...).

```
A -> B -> C -> D
```

Carthage recognizes the Cartfiles of all projects and checks out the correct versions of all dependencies. However, during the build step, Project B is built before D and C. This results in a build error in B.

```
error: no such module 
import C
``` 

When I run `carthage update` for Project B only, everything seems to work fine.
Is there a way to make the order of dependencies explicit?

Have the same issue. 
Dependencies look like this:
zm-client-ios ---> zshare ---> zmc-images ---> zmc-system, zmc-testing

zm-client-ios/Cartfile:
```
github "wearezeta/zclient-ziphy" "1.1.0"
github "kompozer/SmoothLineView" "0.1.9"
github "swisspol/GCDWebServer" "3.2.5"
github "kompozer/TSMarkdownParser" "1.0.16"
github "wearezeta/zshare" "master"
```

zshare/Cartfile:
```
github "wearezeta/zmc-images" "master"
```
zmc-images/Cartfile:
```
github "wearezeta/zmc-system" "master"
github "wearezeta/zmc-testing" "master"
```

When I run carthage update  --use-submodules --platform "iOS" in zm-client-ios I got this:
```
*** Fetching zshare
*** Fetching TSMarkdownParser
*** Fetching GCDWebServer
*** Fetching SmoothLineView
*** Fetching zclient-ziphy
*** Fetching zmc-images
*** Fetching zmc-testing
*** Fetching zmc-system
*** Fetching ocmock
*** Checking out GCDWebServer at "bdfe6728ae5faec83c08f807722090abcde2c496"
*** Checking out zmc-testing at "cad86bddcca5fd55df335dc0015711c0b152f297"
*** Checking out zclient-ziphy at "5434f2dd77f50b572d4a7910138459bdb9d9ca57"
*** Checking out zmc-system at "3ede47ff531e5592bb33e8f876ad8de89f337177"
*** Checking out ocmock at "3b003b2e25f90399db8ff1d3286807bf761aaa21"
*** Checking out SmoothLineView at "6526bdd6e4701d5470c3f730065ac84cf5346dc6"
*** Checking out TSMarkdownParser at "8aa160127a84573f5f4c61ce65dd451275cd91f6"
*** Checking out zshare at "471b073c86431c115ae6042fdcc5d8dd15b37a5b"
*** Checking out zmc-images at "09892cef9184ebeb89a586fbb2afab71887db8e4"
*** xcodebuild output can be found in /var/folders/nt/g2pgfj2x1bjg2h9jrr_l7l5r0000gn/T/carthage-xcodebuild.Sf6519.log
*** Building scheme "GCDWebServers (iOS)" in GCDWebServer.xcodeproj
*** Building scheme "SmoothLineView" in SmoothLineView.xcodeproj
*** Building scheme "MarkdownParser" in TSMarkdownParser.xcodeproj
*** Building scheme "OCMock iOS" in OCMock.xcodeproj
*** Building scheme "ziphy" in ziphy.xcodeproj
*** Building scheme "ZMSystem-ios" in ZMSystem.xcodeproj
*** Building scheme "zshare" in zshare.xcodeproj
** BUILD FAILED **


The following build commands failed:
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/RequestProcessing.swift
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/TransportSession.swift
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/AssetPreprocessing.swift
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/Backend.swift
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/Encoders.swift
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Public/ShareExtensionAPI.swift
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/RequestSigning.swift
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/Decoders.swift
	CompileSwiftSources normal x86_64 com.apple.xcode.tools.swift.compiler
(9 failures)
/Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/AssetPreprocessing.swift:10:8: error: no such module 'zimages'
/Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/AssetPreprocessing.swift:10:8: error: no such module 'zimages'
/Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/AssetPreprocessing.swift:10:8: error: no such module 'zimages'
/Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/AssetPreprocessing.swift:10:8: error: no such module 'zimages'
/Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/AssetPreprocessing.swift:10:8: error: no such module 'zimages'
/Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/AssetPreprocessing.swift:10:8: error: no such module 'zimages'
/Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/AssetPreprocessing.swift:10:8: error: no such module 'zimages'
/Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/AssetPreprocessing.swift:10:8: error: no such module 'zimages'
A shell task failed with exit code 65:
** BUILD FAILED **


The following build commands failed:
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/RequestProcessing.swift
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/TransportSession.swift
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/AssetPreprocessing.swift
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/Backend.swift
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/Encoders.swift
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Public/ShareExtensionAPI.swift
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/RequestSigning.swift
	CompileSwift normal x86_64 /Users/ilya/zclient-ios/Carthage/Checkouts/zshare/zshare/Private/Network/Decoders.swift
	CompileSwiftSources normal x86_64 com.apple.xcode.tools.swift.compiler
(9 failures)
```

zm-client-ios/Cartfile.resolved:
```
github "swisspol/GCDWebServer" "bdfe6728ae5faec83c08f807722090abcde2c496"
github "kompozer/SmoothLineView" "6526bdd6e4701d5470c3f730065ac84cf5346dc6"
github "kompozer/TSMarkdownParser" "8aa160127a84573f5f4c61ce65dd451275cd91f6"
github "wearezeta/ocmock" "3b003b2e25f90399db8ff1d3286807bf761aaa21"
github "wearezeta/zclient-ziphy" "5434f2dd77f50b572d4a7910138459bdb9d9ca57"
github "wearezeta/zmc-system" "3ede47ff531e5592bb33e8f876ad8de89f337177"
github "wearezeta/zshare" "471b073c86431c115ae6042fdcc5d8dd15b37a5b"
github "wearezeta/zmc-testing" "cad86bddcca5fd55df335dc0015711c0b152f297"
github "wearezeta/zmc-images" "09892cef9184ebeb89a586fbb2afab71887db8e4"
```

If I move zshare line in Cartfile.resolved after zmc-images and run ```carthage build --platform "iOS"``` build works. So the problem is in the order of checkout as I can tell.
After changing lines in Carfile.resolved output looks like this:
```
*** Building scheme "GCDWebServers (iOS)" in GCDWebServer.xcodeproj
*** Building scheme "SmoothLineView" in SmoothLineView.xcodeproj
*** Building scheme "MarkdownParser" in TSMarkdownParser.xcodeproj
*** Building scheme "OCMock iOS" in OCMock.xcodeproj
*** Building scheme "ziphy" in ziphy.xcodeproj
*** Building scheme "ZMSystem-ios" in ZMSystem.xcodeproj
*** Building scheme "ZMTesting-ios" in ZMTesting.xcodeproj
*** Building scheme "zimages-ios" in zimages.xcodeproj
*** Building scheme "zshare" in zshare.xcodeproj
```

--- a/Source/CarthageKit/Resolver.swift
+++ b/Source/CarthageKit/Resolver.swift
@@ -201,7 +201,7 @@ private struct DependencyGraph: Equatable {
 	var allNodes: Set<DependencyNode> = []
 
 	/// All nodes that have dependencies, associated with those lists of
-	/// dependencies themselves.
+	/// dependencies themselves including the intermediates.
 	var edges: [DependencyNode: Set<DependencyNode>] = [:]
 
 	/// The root nodes of the graph (i.e., those dependencies that are listed
@@ -280,6 +280,15 @@ private struct DependencyGraph: Equatable {
 			var nodeSet = edges[dependencyOf] ?? Set()
 			nodeSet.insert(node)
 			edges[dependencyOf] = nodeSet
+
+			// Add a nested dependency to the list of its ancestor.
+			let edgesCopy = edges
+			for (ancestor, var itsDependencies) in edgesCopy {
+				if itsDependencies.contains(dependencyOf) {
+					itsDependencies.insert(node)
+					edges[ancestor] = itsDependencies
+				}
+			}
 		} else {
 			roots.insert(node)
 		}

--- a/Source/CarthageKit/Resolver.swift
+++ b/Source/CarthageKit/Resolver.swift
@@ -279,6 +279,12 @@ private struct DependencyGraph: Equatable {
 		if let dependencyOf = dependencyOf {
 			var nodeSet = edges[dependencyOf] ?? Set()
 			nodeSet.insert(node)
+
+			// If the given node has its dependencies, add thme also to the list.
+			if let dependenciesOfNode = edges[node] {
+				nodeSet.unionInPlace(dependenciesOfNode)
+			}
+
 			edges[dependencyOf] = nodeSet
 
 			// Add a nested dependency to the list of its ancestor.

--- a/Source/CarthageKit/Resolver.swift
+++ b/Source/CarthageKit/Resolver.swift
@@ -214,17 +214,21 @@ private struct DependencyGraph: Equatable {
 			let lhsDependencies = self.edges[lhs]
 			let rhsDependencies = self.edges[rhs]
 
-			// If the right node or its nested dependencies have a dependency on
-			// the left node, the left node needs to be built first (and is thus
-			// ordered first).
-			if let rhsDependencies = rhsDependencies where self.edgeDependsOn(rhsDependencies, dependsOn: lhs) {
-				return true
+			if let rhsDependencies = rhsDependencies {
+				// If the right node has a dependency on the left node, the
+				// left node needs to be built first (and is thus ordered
+				// first).
+				if rhsDependencies.contains(lhs) {
+					return true
+				}
 			}
 
-			// If the left node or its nested dependencies have a dependency on
-			// the right node, the right node needs to be built first.
-			if let lhsDependencies = lhsDependencies where self.edgeDependsOn(lhsDependencies, dependsOn: rhs) {
-				return false
+			if let lhsDependencies = lhsDependencies {
+				// If the left node has a dependency on the right node, the
+				// right node needs to be built first.
+				if lhsDependencies.contains(rhs) {
+					return false
+				}
 			}
 
 			// If neither node depends on each other, sort the one with the
@@ -244,22 +248,6 @@ private struct DependencyGraph: Equatable {
 	}
 
 	init() {}
-
-	/// Determines whether the set of nodes or the dependencies of them have a
-	/// dependency on the given node.
-	func edgeDependsOn(edge: Set<DependencyNode>, dependsOn: DependencyNode) -> Bool {
-		if edge.contains(dependsOn) {
-			return true
-		}
-
-		for node in edge {
-			if let nodeDependencies = self.edges[node] where edgeDependsOn(nodeDependencies, dependsOn: dependsOn) {
-				return true
-			}
-		}
-
-		return false
-	}
 
 	/// Attempts to add the given node to the graph, optionally as a dependency
 	/// of another.

--- a/Source/CarthageKit/Resolver.swift
+++ b/Source/CarthageKit/Resolver.swift
@@ -214,21 +214,17 @@ private struct DependencyGraph: Equatable {
 			let lhsDependencies = self.edges[lhs]
 			let rhsDependencies = self.edges[rhs]
 
-			if let rhsDependencies = rhsDependencies {
-				// If the right node has a dependency on the left node, the
-				// left node needs to be built first (and is thus ordered
-				// first).
-				if rhsDependencies.contains(lhs) {
-					return true
-				}
+			// If the right node or its nested dependencies have a dependency on
+			// the left node, the left node needs to be built first (and is thus
+			// ordered first).
+			if let rhsDependencies = rhsDependencies where self.edgeDependsOn(rhsDependencies, dependsOn: lhs) {
+				return true
 			}
 
-			if let lhsDependencies = lhsDependencies {
-				// If the left node has a dependency on the right node, the
-				// right node needs to be built first.
-				if lhsDependencies.contains(rhs) {
-					return false
-				}
+			// If the left node or its nested dependencies have a dependency on
+			// the right node, the right node needs to be built first.
+			if let lhsDependencies = lhsDependencies where self.edgeDependsOn(lhsDependencies, dependsOn: rhs) {
+				return false
 			}
 
 			// If neither node depends on each other, sort the one with the
@@ -248,6 +244,22 @@ private struct DependencyGraph: Equatable {
 	}
 
 	init() {}
+
+	/// Determines whether the set of nodes or the dependencies of them have a
+	/// dependency on the given node.
+	func edgeDependsOn(edge: Set<DependencyNode>, dependsOn: DependencyNode) -> Bool {
+		if edge.contains(dependsOn) {
+			return true
+		}
+
+		for node in edge {
+			if let nodeDependencies = self.edges[node] where edgeDependsOn(nodeDependencies, dependsOn: dependsOn) {
+				return true
+			}
+		}
+
+		return false
+	}
 
 	/// Attempts to add the given node to the graph, optionally as a dependency
 	/// of another.

--- a/Source/CarthageKit/Resolver.swift
+++ b/Source/CarthageKit/Resolver.swift
@@ -280,7 +280,7 @@ private struct DependencyGraph: Equatable {
 			var nodeSet = edges[dependencyOf] ?? Set()
 			nodeSet.insert(node)
 
-			// If the given node has its dependencies, add thme also to the list.
+			// If the given node has its dependencies, add them also to the list.
 			if let dependenciesOfNode = edges[node] {
 				nodeSet.unionInPlace(dependenciesOfNode)
 			}

